vm/proxyapp/proxyappclient_test.go

```diff
@@ -0,0 +1,535 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package proxyapp
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/rpc"
+	"net/rpc/jsonrpc"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/vm/proxyapp/proxyrpc"
+	"github.com/google/syzkaller/vm/vmimpl"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+var testEnv = &vmimpl.Env{
+	Config: []byte(`
+{
+    "cmd": "/path/to/proxyapp_binary",
+    "config": {
+			"internal_values": 123
+    }
+  }
+`)}
+
+func makeTestParams() *proxyAppParams {
+	return &proxyAppParams{
+		CommandRunner:  osutilCommandContext,
+		InitRetryDelay: 0,
+	}
+}
+
+func makeMockProxyAppProcess() (
+	*mock.Mock, io.WriteCloser, io.ReadCloser, io.ReadCloser) {
+	rStdin, wStdin := io.Pipe()
+	rStdout, wStdout := io.Pipe()
+	rStderr, wStderr := io.Pipe()
+	wStderr.Close()
+
+	server := rpc.NewServer()
+	handler := makeMockProxyApp()
+	server.RegisterName("ProxyVM", handler)
+
+	go server.ServeCodec(jsonrpc.NewServerCodec(stdInOutCloser{
+		rStdin,
+		wStdout,
+	}))
+
+	return &handler.Mock, wStdin, rStdout, rStderr
+}
+
+type nopWriteCloser struct {
+	io.Writer
+}
+
+func (nopWriteCloser) Close() error {
+	return nil
+}
+
+func TestCtor_Ok(t *testing.T) {
+	mProxyAppServer, mCmdRunner, params := proxyAppServerFixture()
+	p, err := ctor(params, testEnv)
+
+	assert.Nil(t, err)
+	assert.Equal(t, 2, p.Count())
+
+	<-mCmdRunner.onWaitCalled
+	mProxyAppServer.AssertExpectations(t)
+	mCmdRunner.AssertExpectations(t)
+}
+
+func TestCtor_ReadBadConfig(t *testing.T) {
+	pool, err := ctor(makeTestParams(), &vmimpl.Env{
+		Config: []byte(`{"wrong_key": 1}`),
+	})
+	assert.NotNil(t, err)
+	assert.Nil(t, pool)
+}
+
+func TestCtor_FailedPipes(t *testing.T) {
+	mCmdRunner, params := makeMockCommandRunner()
+	mCmdRunner.
+		On("StdinPipe").
+		Return(nil, fmt.Errorf("stdinpipe error")).
+		Once().
+		On("StdinPipe").
+		Return(nopWriteCloser{&bytes.Buffer{}}, nil).
+		On("StdoutPipe").
+		Return(nil, fmt.Errorf("stdoutpipe error")).
+		Once().
+		On("StdoutPipe").
+		Return(io.NopCloser(strings.NewReader("")), nil).
+		On("StderrPipe").
+		Return(nil, fmt.Errorf("stderrpipe error"))
+
+	for i := 0; i < 3; i++ {
+		p, err := ctor(params, testEnv)
+		assert.NotNil(t, err)
+		assert.Nil(t, p)
+	}
+}
+
+func TestClose_waitDone(t *testing.T) {
+	mProxyAppServer, mCmdRunner, params := proxyAppServerFixture()
+	mCmdRunner.
+		On("waitDone").
+		Return(nil)
+
+	p, _ := ctor(params, testEnv)
+	p.(*pool).Close()
+
+	mProxyAppServer.AssertExpectations(t)
+	mCmdRunner.AssertExpectations(t)
+}
+
+func TestClose_noMoreCalls(t *testing.T) {
+	mProxyAppServer, mCmdRunner, params := proxyAppServerFixture()
+	mCmdRunner.
+		On("waitDone").
+		Return(nil)
+
+	p, _ := ctor(params, testEnv)
+	p.(*pool).Close()
+
+	preStat := mCmdRunner.totalFuncsCalled
```

Comment:
The following 5 lines should be a part of every test, right? We could put them into some helper function.
But then it may be better/simpler to make the mock pool Close set a flag and all methods will panic if this flag is set. Then we don't need the explicit check.
But then this flag/checking probably belongs to vm.Pool itself. Then it will be checked for all vm implementation and not just in tests.

If we remove totalFuncsCalled, would it be possible to generate the mock using https://github.com/vektra/mockery? The mock code really asks for auto-generation.

Reply:
Use flag instead looks good. Minus test, higher impact, higher code locality.

> But then this flag/checking probably belongs to vm.Pool itself. Then it will be checked for all vm implementation and not just in tests.

I don't think so... We don't want syz-manager to fail-fast.
But agree, it could be a part of some integration tests (vm + proxyapp packages).
And probably we'll benefit having some per-vm-type tests.
It is +1 to the end-to-end testing you mentioned few months/weeks ago. I'll add TODO.

Reply:
To wait for activity after Close() anyway some sleep() is needed.
Change to bool instead of totalCalls counter doesn't seem very beneficial now. Especially because I have only 2 tests with .Close() that may potentially trigger this panic.

Reply:
And I'll check https://github.com/vektra/mockery.
These mocks required some tricks to fit tests, thus I'm a little sceptic regarding automation.

Reply:
We fail-fast on our bugs in the code:
```
syzkaller$ git grep "panic(" *.go | wc -l
1556
```
We don't fail-fast on external services/programs.

Reply:
TODO: check "go generate" and automate mock generation

Reply:
If you don't mind, let's add pool.Close() tests as a separate PR.